pkg/services/meta/container.go

```diff
@@ -0,0 +1,129 @@
+package meta
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"maps"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/mpt"
+	"github.com/nspcc-dev/neo-go/pkg/core/storage"
+	"github.com/nspcc-dev/neo-go/pkg/core/storage/dbconfig"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	objectsdk "github.com/nspcc-dev/neofs-sdk-go/object"
+)
+
+type containerStorage struct {
+	m        sync.RWMutex
+	opsBatch map[string][]byte
+
+	path string
+	mpt  *mpt.Trie
+	db   storage.Store
+}
+
+func (s *containerStorage) drop() error {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	err := s.db.Close()
+	if err != nil {
+		return fmt.Errorf("close container storage: %w", err)
+	}
+
+	err = os.RemoveAll(s.path)
+	if err != nil {
+		return fmt.Errorf("remove container storage: %w", err)
+	}
+
+	return nil
+}
+
+func (s *containerStorage) putObject(e objEvent) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	newKVs := make(map[string][]byte)
+	commsuffix := e.oID[:]
+
+	// batching that is implemented for MPT ignores key's first byte
+
+	newKVs[string(append([]byte{0, oidIndex}, commsuffix...))] = []byte{}
+	newKVs[string(append([]byte{0, sizeIndex}, commsuffix...))] = e.size.Bytes()
+	if len(e.firstObject) == sha256.Size {
+		newKVs[string(append([]byte{0, firstPartIndex}, commsuffix...))] = e.firstObject
+	}
+	if len(e.prevObject) == sha256.Size {
+		newKVs[string(append([]byte{0, previousPartIndex}, commsuffix...))] = e.prevObject
+	}
+	if len(e.deletedObjects) > 0 {
+		newKVs[string(append([]byte{0, deletedIndex}, commsuffix...))] = e.deletedObjects
+	}
+	if len(e.lockedObjects) > 0 {
+		newKVs[string(append([]byte{0, lockedIndex}, commsuffix...))] = e.lockedObjects
+	}
+	if e.typ != objectsdk.TypeRegular {
+		newKVs[string(append([]byte{0, typeIndex}, commsuffix...))] = []byte{byte(e.typ)}
+	}
+
+	if s.opsBatch == nil {
+		s.opsBatch = make(map[string][]byte)
+	}
+	maps.Copy(s.opsBatch, newKVs)
+
+	err := s.db.PutChangeSet(mptToStoreBatch(newKVs), nil)
+	if err != nil {
+		return fmt.Errorf("put MPT KVs to the raw storage manually: %w", err)
+	}
+
+	// TODO: receive full object header and store its non-MPT parts in
+	//  persistent object storage
+
+	return nil
+}
+
+// mptToStoreBatch drops the first byte from every key in the map.
+func mptToStoreBatch(b map[string][]byte) map[string][]byte {
+	for k, v := range b {
+		b[k[1:]] = v
```

Comment:
this is unsafe. https://go.dev/ref/spec#For_range states
> If a map entry is created during iteration, that entry may be produced during the iteration or may be skipped.

and u dont drop `k` elements. It's easier to fill a new map in general. But since u have fixed-len keys, this is doable:
```go
	for k, v := range b {
		if len(k) == 34 {
			b[k[1:]] = v
		}
	}
	maps.DeleteFunc(b, func(k string, _ []byte) bool { return len(k) == 34 })
```

since it's very specific, i'd inline this into the only usage place

Reply:
tbh, i did smth strange here, thought about smth different. this is a total mistake, fixed

Reply:
func is still the same, but correct now

Reply:
now it is used twice